src/GraphQL/GraphQLExtensions.cs

```diff
@@ -528,72 +579,26 @@ public static IValue AstFromValue(this object value, ISchema schema, IGraphType
                 {
                     var values = list
                         .Cast<object>()
-                        .Select(item => AstFromValue(item, schema, itemType))
+                        .Select(item => ToAST(itemType, item))
                         .ToList();
 
                     return new ListValue(values);
                 }
 
-                return AstFromValue(value, schema, itemType);
+                return ToAST(itemType, value);
             }
 
             // Populate the fields of the input object by creating ASTs from each value
             // in the dictionary according to the fields in the input type.
             if (type is IInputObjectGraphType input)
             {
-                if (!(value is Dictionary<string, object> dict))
-                {
-                    return null;
-                }
-
-                var fields = dict
-                    .Select(pair =>
-                    {
-                        var fieldType = input.GetField(pair.Key)?.ResolvedType;
-                        return new ObjectField(pair.Key, AstFromValue(pair.Value, schema, fieldType));
-                    })
-                    .ToList();
-
-                return new ObjectValue(fields);
+                return input.ToAst(value) ?? throw new InvalidOperationException($"Unable to convert the '{value}' of the object type '{input.Name}' to an AST representation.");
             }
 
-            if (!(type is ScalarGraphType inputType))
+            if (!(type is ScalarGraphType scalar))
                 throw new ArgumentOutOfRangeException(nameof(type), $"Must provide Input Type, cannot use: {type}");
```

Comment:
```suggestion
                throw new ArgumentOutOfRangeException(nameof(type), $"Must provide Input Type, cannot use '{type}'");
```

Also please invert condition:
```
if (type is ScalarGraphType scalar)
  return scalar.ToAST(value) ?? throw new InvalidOperationException($"Unable to convert '{value}' of the scalar type '{scalar.Name}' to an AST representation.");
        
throw new ArgumentOutOfRangeException(nameof(type), $"Must provide Input Type, cannot use: {type}");
```